File: src/main/java/br/com/dusty/dkits/util/string/StringUtils.java
package br.com.dusty.dkits.util.string;

import br.com.dusty.dkits.Main;
import br.com.dusty.dkits.util.text.Text;
import org.joda.time.Period;
import org.joda.time.format.PeriodFormatter;
import org.joda.time.format.PeriodFormatterBuilder;

import java.util.ArrayList;

public class StringUtils {
	
	private static final char[] ALPHANUMERIC = "ABCDEFGHIKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789".toCharArray();
	
	/**
	 * Divide uma {@link String} em outras (não-formatadas) de tamanho máximo 'max', inserindo-as, em ordem, em uma {@link ArrayList}.
	 *
	 * @param s
	 * @param max
	 * @return
	 */
	public static ArrayList<String> fancySplit(String s, int max) {
		ArrayList<String> arrayList = new ArrayList<>();
		
		String[] fragments = Text.clearFormatting(s).split(" ");
		for(int i = 0; i < fragments.length; i++){
			StringBuilder fragment = new StringBuilder();
			fragment.append(fragments[i]);
			
			while(i + 1 < fragments.length && (fragment.length() + 1 + fragments[i + 1].length()) <= max){
				fragment.append(" ");
				fragment.append(fragments[i + 1]);
				i++;
			}
			
			arrayList.add(fragment.toString());
		}
		
		return arrayList;
	}
	
	/**
	 * Gera uma {@link String} alfanumérica aleatória com diferenciação entre maiúsculas e minúsculas.
	 *
	 * @param length
	 * @return
	 */
	public static String randomString(int length) {
		StringBuilder stringBuilder = new StringBuilder(length);
		
		for(int i = 0; i <= length - 1; i++){
			int index = Main.RANDOM.nextInt(ALPHANUMERIC.length);
			stringBuilder.append(ALPHANUMERIC[index]);
		}
		
		return stringBuilder.toString();
	}
	
	/**
	 * Cria uma {@link String} que indica uma duração de tempo com dias, horas, minutos e segundos.
	 *
	 * @param millis
	 * @return
	 */
	public static String period(long millis) {
		PeriodFormatter periodFormatter = new PeriodFormatterBuilder().appendDays()
		                                                              .appendSuffix("d")
		                                                              .appendSeparator(" ")
		                                                              .appendHours()
		                                                              .appendSuffix("h")
		                                                              .appendSeparator(" ")
		                                                              .appendMinutes()
		                                                              .appendSuffix("min")
		                                                              .appendSeparator(" ")
		                                                              .appendSeconds()
		                                                              .appendSuffix("s")
		                                                              .toFormatter();
		
		Period period = new Period(millis);
		
		return periodFormatter.print(period.toPeriod());
	}
	
	/**
	 * Cria uma {@link ArrayList} contendo todas as {@link String}'s do parâmetro 'arrayList' que iniciarem com o parâmetro 'start'.
	 *
	 * @param start
	 * @param arrayList
	 * @return
	 */
	public static ArrayList<String> sortOut(String start, ArrayList<String> arrayList) {
		for(String s : arrayList){
			if(!s.toLowerCase().startsWith(start.toLowerCase())){
				arrayList.remove(s);
			}
		}
		
		arrayList.sort(String::compareToIgnoreCase);
		
		return arrayList;
	}
}

File: src/main/kotlin/br/com/dusty/dkits/util/StringUtils.kt
package br.com.dusty.dkits.util

import br.com.dusty.dkits.Main
import br.com.dusty.dkits.util.text.Text
import org.joda.time.Period
import org.joda.time.format.PeriodFormatterBuilder
import java.util.ArrayList
import kotlin.Comparator

/**
 * Divide uma [String] em outras (não-formatadas) de tamanho máximo 'max', inserindo-as, em ordem, em uma [ArrayList].
 *
 * @param s
 * @param max
 * @return
 */
fun String.fancySplit(max: Int): ArrayList<String> {
	val arrayList = ArrayList<String>()

	val fragments = Text.clearFormatting(this).split(" ")
	var i = 0
	while (i < fragments.size) {
		val fragment = StringBuilder()
		fragment.append(fragments[i])

		while (i + 1 < fragments.size && fragment.length + 1 + fragments[i + 1].length <= max) {
			fragment.append(" ")
			fragment.append(fragments[i + 1])
			i++
		}

		arrayList.add(fragment.toString())
		i++
	}

	return arrayList
}

/**
 * Formata uma [Collection] de [String] por extenso, separadas por vírgulas.
 *
 * @param collection
 * @return
 */
fun Collection<String>.format(): String {
	val stringBuilder = StringBuilder()

	val iterator = this.iterator()

	if (iterator.hasNext()) stringBuilder.append(iterator.next())

	while (iterator.hasNext()) stringBuilder.append(", ").append(iterator.next())

	return stringBuilder.toString()
}

/**
 * Cria uma [String] que indica uma duração de tempo com dias, horas, minutos e segundos.
 *
 * @param millis
 * @return
 */
fun Long.periodString(): String {
	val periodFormatter = PeriodFormatterBuilder().appendDays().appendSuffix("d").appendSeparator(" ").appendHours().appendSuffix("h").appendSeparator(" ").appendMinutes().appendSuffix("min").appendSeparator(
			" ").appendSeconds().appendSuffix("s").toFormatter()

	val period = Period(this)

	return periodFormatter.print(period.toPeriod())
}

/**
 * Cria uma [ArrayList] contendo todas as [String]'s do parâmetro 'arrayList' que iniciarem com o parâmetro 'start'.
 *
 * @param start
 * @param arrayList
 * @return
 */
fun ArrayList<String>.sortOut(start: String): ArrayList<String> {
	this.filterNot { it.startsWith(start, true) }.forEach { this.remove(it) }

	this.sortWith(Comparator { obj, str -> obj.compareTo(str, ignoreCase = true) })

	return this
}

object StringUtils {

	private val ALPHANUMERIC = "ABCDEFGHIKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789".toCharArray()

	/**
	 * Gera uma [String] alfanumérica aleatória com diferenciação entre maiúsculas e minúsculas.
	 *
	 * @param length
	 * @return
	 */
	fun randomString(length: Int): String {
		val stringBuilder = StringBuilder(length)

		for (i in 0 until length) {
			val index = Main.RANDOM.nextInt(ALPHANUMERIC.size)
			stringBuilder.append(ALPHANUMERIC[index])
		}

		return stringBuilder.toString()
	}
}
